src/Workspaces/Core/Portable/SemanticModelReuse/AbstractSemanticModelReuseLanguageService.cs

```diff
@@ -0,0 +1,96 @@
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SemanticModelReuse
+{
+    internal abstract class AbstractSemanticModelReuseLanguageService<
+        TBaseMethodDeclarationSyntax,
+        TAccessorDeclarationSyntax,
+        TPropertyDeclarationSyntax,
+        TEventDeclarationSyntax> : ISemanticModelReuseLanguageService
+        where TBaseMethodDeclarationSyntax : SyntaxNode
+        where TAccessorDeclarationSyntax : SyntaxNode
+        where TPropertyDeclarationSyntax : SyntaxNode
+        where TEventDeclarationSyntax : SyntaxNode
+    {
+        protected abstract ISyntaxFacts SyntaxFacts { get; }
+
+        public abstract SyntaxNode? TryGetContainingMethodBodyForSpeculation(SyntaxNode node);
+
+        protected abstract Task<SemanticModel?> TryGetSpeculativeSemanticModelWorkerAsync(SemanticModel previousSemanticModel, SyntaxNode currentBodyNode, CancellationToken cancellationToken);
+        protected abstract SyntaxList<TAccessorDeclarationSyntax> GetAccessors(TPropertyDeclarationSyntax property);
+        protected abstract SyntaxList<TAccessorDeclarationSyntax> GetAccessors(TEventDeclarationSyntax @event);
+
+        public Task<SemanticModel?> TryGetSpeculativeSemanticModelAsync(SemanticModel previousSemanticModel, SyntaxNode currentBodyNode, CancellationToken cancellationToken)
+        {
+            var previousSyntaxTree = previousSemanticModel.SyntaxTree;
+            var currentSyntaxTree = currentBodyNode.SyntaxTree;
+
+            // This operation is only valid if top-level equivalent trees were passed in.
+            Contract.ThrowIfFalse(previousSyntaxTree.IsEquivalentTo(currentSyntaxTree, topLevel: true));
+            return TryGetSpeculativeSemanticModelWorkerAsync(previousSemanticModel, currentBodyNode, cancellationToken);
+        }
+
+        protected SyntaxNode GetPreviousBodyNode(SyntaxNode previousRoot, SyntaxNode currentRoot, SyntaxNode currentBodyNode)
+        {
+            if (currentBodyNode is TAccessorDeclarationSyntax currentAccessor)
+            {
+                // in the case of an accessor, have to find the previous accessor in the previous prop/event corresponding
+                // to the current prop/event.
+
+                var currentContainer = currentBodyNode.Ancestors().First(a => a is TEventDeclarationSyntax || a is TPropertyDeclarationSyntax);
+                var previousContainer = GetPreviousBodyNode(previousRoot, currentRoot, currentContainer);
+
+                var currentAccessors = GetAccessors(currentContainer);
+                var previousAccessors = GetAccessors(previousContainer);
+
+                if (currentAccessors.Count != previousAccessors.Count)
+                {
+                    Debug.Fail("Accessor count shouldn't have changed as there were no top level edits.");
+                    return null;
+                }
+
+                return previousAccessors[currentAccessors.IndexOf(currentAccessor)];
```

Comment:
tehre should be no indexing violation.  First, we guarantee we have the same count of accessors.  Second, we're getting the index of currentAccessor within currentAccessors.  The second part *must* succeed.  i.e. it would/should be impossible to have an accessor not be within the accessor list it is contained in.

The debug-asserts are for parts of this code that i don't have full confidence are always certain.  i.e. i'm not 100% certain that if "top level version" stays the same that all these correspondances hold.  But i am 100% certain that an accessor must be in its containing accessor list, or else the wheels have fallen off the wagon entirely.